### ISSUE TITLE
Using optim module with lutorpy

### ISSUE THREAD

**Issue by telmoamaral** [imodpasteur/lutorpy#27]
Hello. I'm trying to replicate the simple optim example given in [https://github.com/torch/optim/blob/master/doc/intro.md](https://github.com/torch/optim/blob/master/doc/intro.md), but using lutorpy:

```python
require('nn')
require('optim')

global gradParams, model, batch_inputs, batch_labels

nr_inputs = 2
nr_hidden = 20
model = nn.Sequential()
model._add(nn.Linear(nr_inputs, nr_hidden))
model._add(nn.Tanh())
model._add(nn.Linear(nr_hidden, 1))

criterion = nn.MSECriterion()

batch_size = 128
batch_inputs = torch.DoubleTensor(batch_size, nr_inputs)
batch_labels = torch.DoubleTensor(batch_size)

for i in range(0, batch_size):
	input_ = torch.randn(2)
	label = None
	if input_[0] * input_[1] > 0:
		label = -1
	else:
		label = +1
	batch_inputs[i] = input_
	batch_labels[i] = label

params, gradParams = model._getParameters()

for epoch in range(0, 50):

	def feval(params):
		
		global gradParams, model, batch_inputs, batch_labels
		
		gradParams._zero()
		
		outputs = model._forward(batch_inputs)
		loss = criterion._forward(outputs, batch_labels)
		dloss_doutputs = criterion._backward(outputs, batch_labels)
		model._backward(batch_inputs, dloss_doutputs)
		params_2, gradParams = model._getParameters()

		return loss, gradParams
	
	optim.sgd(feval, params)
```
Now, this returns error "invalid arguments: DoubleTensor number nil" at `.../lua/5.1/optim/sgd.lua:84`. The reason is actually in line 44 of `sgd.lua`:
```lua
local fx,dfdx = opfunc(x)
```
This line means that, when `feval()` is called, its two return values `loss` and `gradParams` should go to `fx` and `dfdx`, respectively. However, what happens is that **both** values go to `fx` packed as userdata and nothing goes to `dfdx`! Is there anything I can do on the Python side to ensure that, when `feval()` is called from Lua, it will return the two values separately as expected?

The reason I would like this to work is that the optim module allows using things like momentum and weight decay. Many thanks for your time!

**Comment by oeway** [comment]
Hi,

Sorry for the delayed reply, I tried to address the issue (https://github.com/scoder/lupa/issues/83) but failed to get an answer.

I think for now, the easiest way to sort out is you can wrap your training code into functions and put into a .lua file, and then import them with "require(XXX)" in lutorpy.

**Comment by telmoamaral** [comment]
Hi. Many thanks for looking into this. It took me ages to come back to this issue, but now I finally needed an implementation. Your suggestion of using a bit of Lua code works really well. Below I include a working example, in case it could be useful to anyone.

In this example I set not only the learning rate but also the momentum of SGD optimisation, to show that this approach can be used to gain control of extra parameters like momentum and weight decay (as opposed to using the StochasticGradient class, which only lets you set the learning rate).

The Python part:

```python
import lutorpy as lua

def try_optim():

	require("nn")
	require("optim")
	require("call_optim_sgd") # this imports the call_optim_sgd.lua file

	# define model
	nr_inputs = 2
	nr_hidden = 20
	model = nn.Sequential()
	model._add(nn.Linear(nr_inputs, nr_hidden))
	model._add(nn.Tanh())
	model._add(nn.Linear(nr_hidden, 1))

	# flatten model parameters
	params, grad_params = model._getParameters()
	
	# choose training criterion
	criterion = nn.MSECriterion()

	# define sgd optimisation parameters
	optim_state = lua.eval("{learningRate = 0.1, momentum = 0.5}")
 
	# define training set
	batch_size = 128
	batch_inputs = torch.DoubleTensor(batch_size, nr_inputs)
	batch_labels = torch.DoubleTensor(batch_size)
	for i in range(0, batch_size):
		input_ = torch.randn(2)
		label = None
		if input_[0] * input_[1] > 0:
			label = -1
		else:
			label = +1
		batch_inputs[i] = input_
		batch_labels[i] = label

	# train model
	for epoch in range(0, 100):
		loss = call_optim_sgd(model, params, grad_params, criterion, optim_state, batch_inputs, batch_labels)
		print loss[0]

	# use model on a batch of 4 samples
	x = torch.DoubleTensor(4, 2)
	x[0][0] =  0.5; x[0][1] =  0.5
	x[1][0] =  0.5; x[1][1] = -0.5
	x[2][0] = -0.5; x[2][1] =  0.5
	x[3][0] = -0.5; x[3][1] = -0.5
	print(model._forward(x))

if __name__ == "__main__":

	try_optim()
```

And the `call_optim_sgd.lua` file:

```lua
function call_optim_sgd(model, params, gradParams, criterion, optimState, batchInputs, batchLabels)

	function feval(params)
	
		gradParams:zero()

		local outputs = model:forward(batchInputs)
		local loss = criterion:forward(outputs, batchLabels)
		local dloss_doutputs = criterion:backward(outputs, batchLabels)
		model:backward(batchInputs, dloss_doutputs)

		return loss, gradParams
		
	end

	params_new, loss = optim.sgd(feval, params, optimState)
	
	return loss

end
```

**Comment by oeway** [comment]
Thanks for your feedback and glad to hear that worked for you.
